File: crates/hyper_rhi/src/surface.rs
//
// Copyright (c) 2024, SkillerRaptor
//
// SPDX-License-Identifier: MIT
//

use std::{fmt::Debug, sync::Arc};

use downcast_rs::Downcast;
use raw_window_handle::{DisplayHandle, WindowHandle};

use hyper_math::UVec2;

use crate::texture::Texture;

#[derive(Clone, Debug)]
pub struct SurfaceDescriptor<'a> {
    pub display_handle: DisplayHandle<'a>,
    pub window_handle: WindowHandle<'a>,
    pub window_size: UVec2,
}

pub trait Surface: Downcast {
    fn resize(&mut self, width: u32, height: u32);
    fn current_texture(&self) -> Arc<dyn Texture>;
}

downcast_rs::impl_downcast!(Surface);

File: hyper_engine/src/rhi/surface.cpp
/*
 * Copyright (c) 2025-present, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

#include "hyper_rhi/surface.hpp"

namespace hyper_engine
{
    void Surface::resize(const uint32_t width, const uint32_t height)
    {
        if (m_width == width || m_height == height)
        {
            return;
        }

        m_resized = true;
        m_width = width;
        m_height = height;
    }

    bool Surface::resized() const
    {
        return m_resized;
    }

    uint32_t Surface::width() const
    {
        return m_width;
    }

    uint32_t Surface::height() const
    {
        return m_height;
    }
} // namespace hyper_engine